src/psyclone/psyir/tools/dependency_tools.py

```diff
@@ -273,102 +224,443 @@ def _is_loop_suitable_for_parallel(self, loop, only_nested_loops=True):
+        would have a distance of 1: solving i=(i+di)-1 gives di=1
+        Similarly, `a(i) = a(i) + 1` has a distance of 0, and
+        `a(2*i)=a(2*i+1)` will have a distance of -0.5 (which indicates that
+        with an index that's an integer the accesses are independent). More
+        complex examples can have several solutions, e.g. `a(i*i) = a(i*i)+1`.
+        This will have the distance 0 and `-2*i`. The latter would result
+        in a negative loop index (i+(-2*i)<0), indicating that actually no
+        dependency exists.
+        This function will return the distance, if it is an integer value
+        independent of loop iterations, and None otherwise.
+        Multiple solution are not handled yet, since they are unlikely to
+        be observed in code handled with PSyclone, so they will also
+        return None.
+
+        :param str var_name: name of the one variable used in the two \
+            index expressions.
+        :param index_read: the index expression of the variable read that \
+            is to be compare.
+        :type index_read: :py:class:`psyclone.psyir.nodes.Node`
+        :param index_written: the index expression of the variable written \
+            that is to be compare.
+        :type index_written: :py:class:`psyclone.psyir.nodes.Node`
+
+        :returns: the dependency distance in loop iterations if it is a
+            independent integer value, and None otherwise.
+        :rtype: int or None.
+
+        '''
+        sym_maths = SymbolicMaths.get()
+        sympy_expressions, symbol_map = SymPyWriter.\
+            get_sympy_expressions_and_symbol_map([index_read, index_written])
+        # If the subscripts do not even depend on the specified variable,
+        # any dependency distance is possible (e.g. `do i ... a(j)=a(j)+1`)
+        if var_name not in symbol_map:
+            return None
+
+        var = symbol_map[var_name]
+        # Create a unique 'dx' variable name if 'x' is the variable.
+        d_var_name = "d_"+var_name
+        idx = 1
+        while d_var_name in symbol_map:
+            d_var_name = f"d{idx}_{var_name}"
+
+        # Create a sympy symbol for this new variable
+        d_var = Symbol(d_var_name)
+
+        # Replace 'var' with 'var+d_var' in the read expression in order
+        # to determine distance between the two accesses:
+        sympy_expressions[1] = sympy_expressions[1].subs({var: (var+d_var)})
+
+        # Now solve for `d_var` to identify the distance
+        solutions = sym_maths.solve_equal_for(sympy_expressions[0],
+                                              sympy_expressions[1],
+                                              d_var)
+        if solutions == "independent":
+            # The solution is independent of the variable, i.e. every value
+            # of `d_var` is a solution. So there are certainly
+            # dependencies.
+            return None
+
+        if len(solutions) == 1:
```

Comment:
Some Nemo files fail here with "TypeError: object of type 'ConditionSet' has no len()". As I said previously, if possible, it would be great to handle this complex sets inside the solve_equal_for and just return a standard python set.